desc/magnetic_fields.py

```diff
@@ -109,10 +197,187 @@ def compute_magnetic_field(self, coords, params={}, basis="rpz"):
+            If None, uses the default parameters for each field.
+            If a list or tuple, should have one entry for each component field.
+        sym : str, optional
+            if Bnormal is symmetric, by default "sin"
+            NOTE: BNORM code only ever deals with sin-symmetric modes, so results
+            may not be as expected if attempt to create a BNORM file with a
+            non-symmetric Bnormal distribution, as only the sin-symmetric modes
+            will be saved.
+        scale_by_curpol : bool, optional
+            Whether or not to scale the Bnormal coefficients by curpol
+            which is expected by most other codes that accept BNORM files,
+            by default True
+
+        Returns
+        -------
+        None
+        """
+        if sym != "sin":
+            raise UserWarning(
+                "BNORM code assumes that |B| has sin symmetry,"
+                + " and so BNORM file only saves the sin coefficients!"
+                + " Resulting BNORM file will not contain the cos modes"
+            )
+
+        if isinstance(surface, EquilibriaFamily):
+            surface = surface[-1]
+        if isinstance(surface, Equilibrium):
+            eq = surface
+            surface = eq.surface
+        else:
+            eq = None
+        if scale_by_curpol and eq is None:
+            raise RuntimeError(
+                "an Equilibrium must be supplied when scale_by_curpol is True!"
+            )
+        if eval_grid is None:
+            eval_grid = LinearGrid(
+                rho=jnp.array(1.0), M=2 * basis_M, N=2 * basis_N, NFP=surface.NFP
+            )
+
+        basis = DoubleFourierSeries(M=basis_M, N=basis_N, NFP=surface.NFP, sym=sym)
+        trans = Transform(basis=basis, grid=eval_grid, build_pinv=True)
+
+        # compute Bnormal on the grid
+        Bnorm, _ = self.compute_Bnormal(
+            surface, eval_grid=eval_grid, source_grid=source_grid, params=params
+        )
+
+        # fit Bnorm with Fourier Series
+        Bnorm_mn = trans.fit(Bnorm)
+        # convert to VMEC-style mode numbers to conform with BNORM format
+        xm, xn, s, c = ptolemy_identity_rev(
+            basis.modes[:, 1], basis.modes[:, 2], Bnorm_mn.reshape((1, Bnorm_mn.size))
+        )
+
+        Bnorm_xn = -xn  # need to negate Xn for BNORM code format of cos(mu+nv)
+
+        # BNORM also scales values by curpol, a VMEC output which is calculated by
+        # (source:
+        #  https://princetonuniversity.github.io/FOCUS/
+        #   notes/Coil_design_codes_benchmark.html )
+        # "BNORM scales B_n by curpol=(2*pi/nfp)*bsubv(m=0,n=0)
+        # where bsubv is the extrapolation to the last full mesh point of
+        # bsubvmnc."
+        # this corresponds to 2pi/NFP*G(rho=1) in DESC
+        curpol = (
+            (
+                2
+                * jnp.pi
+                / surface.NFP
+                * eq.compute("G", grid=LinearGrid(rho=jnp.array(1)))["G"]
```

Comment:
Don't we need to set `M` and `N` in order to get a good flux surface average? Otherwise it will default to returning $B_\theta$ at $\theta=\zeta=0$.

Reply:
the logic in eq.compute should use the correct M and N for surface averaged quantities